pkg/k8s/apis/cilium.io/v2/validator/unknown_fields.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/jeremywohl/flatten"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// detectUnknownFields will check the given policy against the expected policy
+// for any unknown fields that were found. The expected policy is retrieved by
+// taking the given policy, marshalling it into its Golang type (CNP or CCNP),
+// and unmarshalling it back into bytes. We can rely on this process to strip
+// away "unknown" fields (specifically, fields that don't have a `json:...`
+// annotation), hence it is expected. Once it's in bytes, we convert it to an
+// unstructured.Unstructured type so that it matches with the given policy. A
+// diff is performed between the two to uncover any differences.
+//
+// Special treatment is given if a top-level description field is found, and
+// returns ErrTopLevelDescriptionFound if so. This is likely to be the most
+// common path of this function as we can usually rely on the validation of the
+// CRDs themselves, however the top-level description field has been widely
+// used incorrectly (see https://github.com/cilium/cilium/issues/13155).
+//
+// This function returns the following possible errors:
+//   - ErrTopLevelDescriptionFound
+//   - ErrUnknownFields
+//   - Other errors
+func detectUnknownFields(policy *unstructured.Unstructured) error {
+	kind := policy.GetKind()
+
+	scopedLog := log
+	switch kind {
+	case cilium_v2.CNPKindDefinition:
+		scopedLog = scopedLog.WithField(logfields.CiliumNetworkPolicyName,
+			policy.GetName())
+	case cilium_v2.CCNPKindDefinition:
+		scopedLog = scopedLog.WithField(logfields.CiliumClusterwideNetworkPolicyName,
+			policy.GetName())
+	default:
+		return fmt.Errorf("unknown kind %q", kind)
+	}
+
+	if _, ok := policy.Object["description"]; ok {
+		scopedLog.Warn(warnTopLevelDescriptionField)
+		return ErrTopLevelDescriptionFound
+	}
```

Comment:
We don't need to specifically check for the `"description"` field. A [C]CNP can only have these top level fields: `apiVersion`, `kind`, `metadata` and `spec` or `specs`. If any key of the `policy.Object` map contains a key that does not belong to any of these fields then we should return an error.

Reply:
Discussed offline. We are keeping this check because we want to have a specialized warning message about top-level description, so that we can advise users to modify their policies, instead of removing the "unknown" fields, which is what we recommend for any other field found.